plugin/pkg/auth/authorizer/node/graph.go

```diff
@@ -0,0 +1,244 @@
+	namespacedNodes[name] = node
+	g.graph.AddNode(node)
+
+	return node
+}
+
+// must be called under write lock
+func (g *Graph) deleteNode(nodeType nodeType, namespace, name string) {
+	node, exists := g.getNode(nodeType, namespace, name)
+	if !exists {
+		return
+	}
+
+	// find existing neighbors
+	neighborsToRemove := []graph.Node{}
+	g.graph.VisitFrom(node, func(neighbor graph.Node) bool {
+		if g.graph.Degree(neighbor) == 1 {
+			neighborsToRemove = append(neighborsToRemove, neighbor)
+		}
+		return true
+	})
+	g.graph.VisitTo(node, func(neighbor graph.Node) bool {
+		if g.graph.Degree(neighbor) == 1 {
+			neighborsToRemove = append(neighborsToRemove, neighbor)
+		}
+		return true
+	})
+
+	// remove the node
+	g.graph.RemoveNode(node)
+	delete(g.nodes[nodeType][namespace], name)
+	if len(g.nodes[nodeType][namespace]) == 0 {
+		delete(g.nodes[nodeType], namespace)
+	}
+	if len(g.nodes[nodeType]) == 0 {
+		delete(g.nodes, nodeType)
+	}
+
+	// remove neighbors that are now edgeless
+	for _, neighbor := range neighborsToRemove {
+		g.graph.RemoveNode(neighbor)
+		n := neighbor.(*NamedNode)
+		delete(g.nodes[n.nodeType][n.namespace], n.name)
+		if len(g.nodes[n.nodeType][n.namespace]) == 0 {
+			delete(g.nodes[n.nodeType], n.namespace)
+		}
+		if len(g.nodes[n.nodeType]) == 0 {
+			delete(g.nodes, n.nodeType)
+		}
+	}
+}
+
+// AddPod sets up edges for the following relationships:
+//   secret -> pod
+//   configmap -> pod
+//   pvc -> pod
+//   pod -> node
+func (g *Graph) AddPod(pod *api.Pod) {
+	g.lock.Lock()
+	defer g.lock.Unlock()
+
+	g.deleteNode(PodNode, pod.Namespace, pod.Name)
+	podNode := g.getOrCreateNode(PodNode, pod.Namespace, pod.Name)
+	nodeNode := g.getOrCreateNode(NodeNode, "", pod.Spec.NodeName)
+	g.graph.SetEdge(NewNodeEdge(podNode, nodeNode, nodeNode))
+
+	podutil.VisitPodSecretNames(pod, func(secret string) bool {
+		g.graph.SetEdge(NewNodeEdge(g.getOrCreateNode(SecretNode, pod.Namespace, secret), podNode, nodeNode))
+		return true
+	})
+
+	podutil.VisitPodConfigmapNames(pod, func(configmap string) bool {
+		g.graph.SetEdge(NewNodeEdge(g.getOrCreateNode(ConfigMapNode, pod.Namespace, configmap), podNode, nodeNode))
+		return true
+	})
+
+	for _, v := range pod.Spec.Volumes {
+		if v.PersistentVolumeClaim != nil {
+			g.graph.SetEdge(NewNodeEdge(g.getOrCreateNode(PVCNode, pod.Namespace, v.PersistentVolumeClaim.ClaimName), podNode, nodeNode))
+		}
+	}
+}
+func (g *Graph) DeletePod(name, namespace string) {
+	g.lock.Lock()
+	defer g.lock.Unlock()
+	g.deleteNode(PodNode, namespace, name)
+}
+
+// AddPV sets up edges for the following relationships:
+//   secret -> pv
+//   pv -> pvc
+func (g *Graph) AddPV(pv *api.PersistentVolume) {
+	g.lock.Lock()
+	defer g.lock.Unlock()
+
+	g.deleteNode(PVNode, "", pv.Name)
+	pvNode := g.getOrCreateNode(PVNode, "", pv.Name)
+
+	if pv.Spec.ClaimRef != nil {
+		g.graph.SetEdge(simple.Edge{F: pvNode, T: g.getOrCreateNode(PVCNode, pv.Spec.ClaimRef.Namespace, pv.Spec.ClaimRef.Name)})
```

Comment:
Why is this edge different?  It's not immediately obvious to me.  I'm assuming its related to the fact that the PV doesn't directly reference a kubernetesNode?  The asymmetry takes more thought than I'd like.  Comment please.